File: 590017213_PremShandilya_07112024.java
import java.util.Scanner;

public class CustomHashTable {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int table[] = new int[10];
        String input = "";

        while (!input.equals("exit")) {
            System.out.println("Enter the number of element to be inserted or type 'exit' to stop");
            input = scanner.nextLine();
            if (!input.equals("exit")) {
                table = addElement(table, input);
            }
        }

        System.out.println("Enter value to search");
        String searchKey = scanner.nextLine();
        findElement(table, searchKey);

        for (int i = 0; i < table.length; i++) {
            System.out.print(table[i] + " ");
        }

        scanner.close();
    }

    public static int[] addElement(int table[], String input) {
        int hashValue = 0;
        for (int i = 0; i < input.length(); i++) {
            hashValue += (int) input.charAt(i);
        }
        int index = hashValue % table.length;
        boolean added = false;

        if (table[index] == 0) {
            table[index] = hashValue;
        } else {
            for (int i = index + 1; i < table.length; i++) {
                if (table[i] == 0) {
                    table[i] = hashValue;
                    added = true;
                    break;
                }
            }
            if (!added) {
                System.out.println("Hash table is full");
            }
        }
        return table;
    }

    public static void findElement(int table[], String searchKey) {
        int hashValue = 0;
        for (int i = 0; i < searchKey.length(); i++) {
            hashValue += (int) searchKey.charAt(i);
        }
        int index = hashValue % table.length;
        boolean located = false;

        if (table[index] == hashValue) {
            System.out.println("Element found at position " + index);
        } else {
            for (int i = index + 1; i < table.length; i++) {
                if (table[i] == hashValue) {
                    System.out.println("Element found at position " + i);
                    located = true;
                    break;
                }
            }
            if (!located) {
                System.out.println("Element not found");
            }
        }
    }
}

File: 590017213_PremShandilya_07112024.py
class HashTable:
    def __init__(self, size=10):
        self.size = size
        self.table = [None] * self.size

    def hash_function(self, key):
        return key % self.size

    def insert(self, key):
        index = self.hash_function(key)

        original_index = index
        while self.table[index] is not None:
            if self.table[index] == key:
                print(f"Key {key} already exists at index {index}.")
                return
            
            index = (index + 1) % self.size
            if index == original_index:
                print("Hash table is full!")
                return
        
        self.table[index] = key
        print(f"Inserted key {key} at index {index}.")

    def search(self, key):
        index = self.hash_function(key)

        original_index = index
        while self.table[index] is not None:
            if self.table[index] == key:
                print(f"Key {key} found at index {index}.")
                return index
            
            index = (index + 1) % self.size
            if index == original_index:
                break
        
        print(f"Key {key} not found.")
        return None

hash_table = HashTable()

hash_table.insert(12)
hash_table.insert(24)
hash_table.insert(36)
hash_table.insert(42)


hash_table.search(24)
hash_table.search(36)
